### PR TITLE
Improve data processing to enable downloading LAOIN 400M

### PR DESCRIPTION
**Pull request by tchaton**
## What does this PR do?

This PR makes several improvements to the serialization, the parquer reader made to enable us to download LAOIN 400M on the lightning.ai platform.

Fixes #\<issue_number>

<details>
  <summary><b>Before submitting</b></summary>

- Was this **discussed/agreed** via a GitHub issue? (not for typos and docs)
- [ ] Did you read the [contributor guideline](https://github.com/Lightning-AI/lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [ ] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- Did you make sure to **update the documentation** with your changes? (if necessary)
- Did you write any **new necessary tests**? (not for typos and docs)
- [ ] Did you verify new and **existing tests pass** locally with your changes?
- Did you list all the **breaking changes** introduced by this pull request?
- Did you **update the CHANGELOG**? (not for typos, docs, test updates, or minor internal changes/refactors)

</details>

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing, make sure you have read the [review guidelines](https://github.com/Lightning-AI/lightning/wiki/Review-guidelines). In short, see the following bullet-list:

<details>
  <summary>Reviewer checklist</summary>

- [ ] Is this pull request ready for review? (if not, please submit in draft mode)
- [ ] Check that all items from **Before submitting** are resolved
- [ ] Make sure the title is self-explanatory and the description concisely explains the PR
- [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified

</details>

----
📚 Documentation preview 📚: https://pytorch-lightning--19452.org.readthedocs.build/en/19452/

cc @borda

### CODE REVIEW

**Review by dmitsf — APPROVED** [review]

**Review by ethanwharris — APPROVED** [review]

File: src/lightning/data/processing/data_processor.py [hunk 1]
```diff
@@ -713,14 +713,16 @@ def _done(self, size: int, delete_cached_files: bool, output_dir: Dir) -> _Resul
             size = sum([c["dim"] if c["dim"] is not None else c["chunk_size"] for c in config["chunks"]])
             num_bytes = sum([c["chunk_bytes"] for c in config["chunks"]])
             data_format = tree_unflatten(config["config"]["data_format"], treespec_loads(config["config"]["data_spec"]))
+            num_chunks = len(config["chunks"])
+            num_bytes_per_chunk = [c["chunk_size"] for c in config["chunks"]] if num_chunks < 1024 else []
```
Comment (awaelchli) [inline src/lightning/data/processing/data_processor.py:717]: Might be good to put a comment for the hardcoded 1024 limit

File: src/lightning/data/__init__.py [hunk 2]
```diff
@@ -1,3 +1,4 @@
+from lightning.data.constants import RequirementCache
```
Comment (awaelchli) [inline src/lightning/data/__init__.py:1]: Shouldn't this be imported from lightning_utilities?

File: src/lightning/data/processing/utilities.py [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+
+    with urllib.request.urlopen(urllib.request.Request(url, data=None, headers={"User-Agent": user_agent_string}), timeout=timeout) as r:  # noqa: E501, S310
+        img_stream = io.BytesIO(r.read())
+    return img_stream
+
+
+@contextmanager
+def optimize_dns_context(enable: bool) -> Any:
+    optimize_dns(enable)
+    try:
```
Comment (awaelchli) [inline src/lightning/data/processing/utilities.py:44]: shouldn't it be

```
yield
try
...
except
...
```

**Review by awaelchli — APPROVED** [review]

File: src/lightning/data/processing/readers.py [hunk 4]
```diff
@@ -32,100 +31,66 @@ def read(self, item: Any) -> Any:
...
+    def _get_num_rows(self, path: str) -> int:
+        df = ds.dataset(path).scanner()
+        return df.count_rows()
 
-            df = df.take([item.start, item.end])
+    def read(self, filepath: str) -> Any:
+        # Try to force dellocation to avoid memory leak
+        try:
+            import pyarrow as pa
+            pa.jemalloc_set_decay_ms(0)
```
[the remark this answers is not in the document]
Comment (tchaton) [reply src/lightning/data/processing/readers.py:55]: done !

File: src/lightning/data/processing/data_processor.py (hunk 1 above)
Comment (tchaton) [reply src/lightning/data/processing/data_processor.py:717]: done.

File: src/lightning/data/__init__.py (hunk 2 above)
Comment (tchaton) [reply src/lightning/data/__init__.py:1]: done !

File: src/lightning/data/processing/utilities.py (hunk 3 above)
Comment (tchaton) [reply src/lightning/data/processing/utilities.py:44]: No, we want to catch any failures to disable back the context.

**Pull request merged**